kolena/_experimental/data_ingestion/_data_format/coco.py

```diff
@@ -0,0 +1,255 @@
+        return {int(record["id"]): record for record in raw_data["images"]}
+
+    def _get_label_map(self, raw_data):
+        return {int(record["id"]): record["name"] for record in raw_data["categories"]}
+
+    def _get_inf_map(self, raw_inferences, image_map, label_map):
+        ids = set(image_map.keys())
+        image_to_boxes: Dict[int, List[ScoredLabeledBoundingBox]] = defaultdict(lambda: [])
+        for inference in raw_inferences:
+            image_id = inference["image_id"]
+            label = label_map[inference["category_id"]]
+
+            # check that the box is in a valid image
+            if image_id in ids:
+                bbox = inference["bbox"]
+                top_left = (bbox[0], bbox[1])
+                bottom_right = (bbox[0] + bbox[2], bbox[1] + bbox[3])
+                score = inference["score"]
+                bounding_box = ScoredLabeledBoundingBox(top_left, bottom_right, label, score)
+                image_to_boxes[image_id].append(bounding_box)
+        return image_to_boxes
+
+    def _setup_evaluator(self) -> ObjectDetectionEvaluator:
+        return ObjectDetectionEvaluator(
+            configurations=[
+                ThresholdConfiguration(
+                    threshold_strategy=0.3,
+                    iou_threshold=0.3,
+                    with_class_level_metrics=True,
+                    min_confidence_score=0.2,
+                ),
+                ThresholdConfiguration(
+                    threshold_strategy=F1_OPTIMAL,
+                    iou_threshold=0.5,
+                    with_class_level_metrics=True,
+                    min_confidence_score=0.0,
+                ),
+            ],
+        )
+
+    def _save_data(self) -> None:
+        model_name = self.config.model_name
+        test_suite_name = self.config.test_suite_name
+        reset = self.config.reset
+        # create a model
+        model = Model(model_name, infer=self._get_infer(self.inf_map))
+
+        # customizable configurations for the evaluator
+        evaluator = self._setup_evaluator()
+
+        # runs the evaluation
+        test_suite = TestSuite(test_suite_name)
+        test(model, test_suite, evaluator, reset=reset)
+
+    def _get_image_id(self, locator: str) -> int:
+        """
+        e.g. s3://kolena-public-datasets/coco-2014-val/imgs/COCO_val2014_000000268396.jpg => 268396
+        """
+        filename = locator.split("/")[-1]
+        image_id = filename.split(".")[0].split("_")[-1]
+        return int(image_id)
+
+    def _get_infer(
+        self,
+        inf_map: Dict[int, List[ScoredLabeledBoundingBox]],
+    ) -> Callable[[TestSample], Inference]:
+        # a function that creates an inference from a test sample
+        def infer(sample: TestSample) -> Inference:
+            try:
+                locator = sample.locator
+                # TODO: maybe we should save the id for test samples
+                image_id = self._get_image_id(locator)
```

Comment:
this is fragile, the locator that derived from `file_name` might not match to image id that is used in the inference coco json, we should to add id field in the pre-build test samples, but it might be a breaking change

Reply:
Definitely very fragile -- I'd be a fan of saving lookup information in the sample itself, to help with resolution. Can we not declare an extended version of `TestSample` that includes `coco_image_id` as a field? I thought we were able to get extensions of the base class to work (with customer data).

Reply:
An alternative is ingesting test suite and results at the same time, to ensure that we're dealing with the same internal ids in the pair of files.

Reply:
thanks for the ideas, seems like [extending the class](https://github.com/kolenaIO/kolena/pull/196/commits/b66528bbe172d35929f19b8c539cffd3105d300d) can fix this issue